src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1714 @@
+
+        if (input_ids is None) ^ (inputs_embeds is not None):
+            raise ValueError(
+                "You cannot specify both input_ids and inputs_embeds at the same time, and must specify either one"
+            )
+
+        if self.gradient_checkpointing and self.training and use_cache:
+            logger.warning_once(
+                "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`."
+            )
+            use_cache = False
+
+        if inputs_embeds is None:
+            inputs_embeds = self.embed_tokens(input_ids)
+
+        hidden_states = inputs_embeds
+        
+        x_orig = torch.clone(inputs_embeds)
+
+
+        if use_cache and past_key_values is None:
+            logger.warning_once(
+                "Zamba requires an initialized `HybridMambaAttentionDynamicCache` to return a cache. None was "
+                "provided, so no cache will be returned."
+            )
+
+        if cache_position is None:
+            cache_position = torch.arange(hidden_states.shape[1], device=hidden_states.device)
+        if position_ids is None:
+            position_ids = cache_position.unsqueeze(0)
+
+        causal_mask = self._update_causal_mask(attention_mask, inputs_embeds, cache_position)
+
+        all_hidden_states = () if output_hidden_states else None
+        all_self_attns = () if output_attentions else None
+
+        if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+            
+        mamba_layers = iter(self.mamba_layers)
+        linear_layers = iter(self.linear_layers)
+
+        
+        
+
+        for layer_id_num, layer_type in enumerate(self.layers_block_type):
+
+            if layer_type == "attention+mamba":
+                if self.gradient_checkpointing and self.training:
+                    layer_outputs = self._gradient_checkpointing_func(
+                        self.block.__call__,
+                        hidden_states,
+                        x_orig,
+                        layer_id_num,
+                        causal_mask,
+                        position_ids,
+                        past_key_values,
+                        output_attentions,
+                        use_cache,
+                        cache_position,
+                    )
+                else:
+                    layer_outputs = self.block(
+                        hidden_states,
+                        x_orig=x_orig,
+                        layer_id_num=layer_id_num,
+                        attention_mask=causal_mask,
+                        position_ids=position_ids,
+                        past_key_value=past_key_values,
+                        output_attentions=output_attentions,
+                        use_cache=use_cache,
+                        cache_position=cache_position,
+                    )
+                from_tf = layer_outputs[0]
```

Comment:
it is unclear to me what `from_tf` is used for

Reply:
As mentioned above, we added a comment to explain the role of from_tf around [this line](https://github.com/huggingface/transformers/blob/aa576ac536e983f5f42a5fa885c4d82736f736d7/src/transformers/models/zamba/modeling_zamba.py#L1047), which reads:

```
from_tf is the output of shared transformer + linear layer (these layers are shown in fig. 2 in https://arxiv.org/pdf/2405.16712).
from_tf is then added to the input to the mamba layer (as described in eq. (6) of https://arxiv.org/pdf/2405.16712, where y_l in that equation is from_tf).
```
Please let us know if anything is unclear. Thank you for helping us improve our code!

Reply:
Thanks ! This is clearer ! my only comment is about making the variable name more self explanatory, see my comment above 🙏